app/src/main/java/app/shoppinglist/wsux/shoppinglist/firebase/ShopList.java

```diff
@@ -133,8 +133,57 @@ void removeTaskFromList(final ShopTask shopTask) {
         }
     }
 
-    public void remove() {
+    public boolean remove() {
+        return removeListAsAuthor() || quitListAsAuthor() || quitListAsCollaborator();
+    }
+
+    private boolean removeListAsAuthor() {
+        if (!userInfo.getUserId().equals(author) || collaborators.size() != 0) {
+            return false;
+        }
+
+        ref.delete().addOnSuccessListener(new OnSuccessListener<Void>() {
+            @Override
+            public void onSuccess(Void aVoid) {
+                setNotReady();
+                removeAllListeners();
+                userInfo.removeKnownList(listId);
+                userInfo.reportChildChange();
+            }
+        }).addOnFailureListener(this);
+
+        return true;
+    }
+
+    private boolean quitListAsAuthor() {
+        if (!userInfo.getUserId().equals(author) || collaborators.size() == 0) {
+            return false;
+        }
+
+        String firstCollaboratorFound = collaborators.get(0);
```

Comment:
I think these lines should be in a different function

private void authorIsFirstCollaborator(){
        String firstCollaboratorFound = collaborators.get(0);
        removeCollaborators(firstCollaboratorFound);
         if (author.equals(firstCollaboratorFound)) {
            return remove();
        }
}